### ISSUE TITLE
passing directly constituency parse to tregex

### ISSUE THREAD

**Issue by valeryweber** [stanfordnlp/python-stanford-corenlp#31]
Dear All
Is it possible to pass directly constituency parse to tregex?
Many thanks
valery

**Comment by tanloong** [comment]
I encountered the same problem. Is there a solution now?

**Comment by AngledLuffa** [comment]
I've been looking for a use case in order to make a tregex interface like that Python.  (I've needed semgrex and tsurgeon for my own work, but never tregex yet.)  It would not take long to make, though.

What kind of interface are you looking for?  Do you need to have the subtrees themselves returned, or is it sufficient to have a boolean yes/no for if the pattern matches a tree?

**Comment by AngledLuffa** [comment]
No, I would make an interface where you could pass in trees and a query to
CoreNLP's implementation of tregex, and you would get back something.  What
that something is I need to figure out - I've never needed anything from
tregex myself except in situations where it was just simpler to use the
Java interface directly

**Comment by tanloong** [comment]
Oh, I get it now!

I am writing a program to count frequencies of a few syntactic structures. Recently I am moving the dependencies of it from the standalone Stanford Parser and Stanford Tregex to CoreNLP. [This](https://pastebin.com/SfkUPPwQ) is an example about how far I have gone with CoreNLP.

What I want:
1. Save the parsed trees to disk.
2. For each tregex pattern, save the matched subtrees to disk.
3. The program should be lightweight, so that installing it would be easy. For this reason, stanza is not what I am looking for, because the language model it requires to install is often unaccessible for my potential users, who are, like me, inside of China.

What I encountered:
1. If I try to save parsed trees, the text has to be parsed twice, which is inefficient.
2. If I want the code to be efficient (only parse the text when querying it), I can't get the complete parsed trees (I can only get matched subtrees).

If we can let tregex handle passed-in trees, my problem would be solved. 😀

**Comment by AngledLuffa** [comment]
Ah, so a Java solution would work for you?

The Stanza Python interface would not need to involve the parser if you have already parsed trees and want a Python interface to something like Tregex

**Comment by tanloong** [comment]
Yes, I think a Java solution works for me if I can send parsed trees, instead of plain text, to `https://localhost:9000/tregex`.

**Comment by AngledLuffa** [comment]
> if I can directly send parsed trees to `https://localhost:9000/tregex`

I think that already exists.  You can send your query with the `serialized` format instead of the `text` format, and it will run the tregex patterns on the existing trees.

You can see here that the `tregex` endpoint is using `getDocument`:

https://github.com/stanfordnlp/CoreNLP/blob/b167dc2dda4bd68cc4a6156e051408e43598f1a2/src/edu/stanford/nlp/pipeline/StanfordCoreNLPServer.java#L1332
https://github.com/stanfordnlp/CoreNLP/blob/b167dc2dda4bd68cc4a6156e051408e43598f1a2/src/edu/stanford/nlp/pipeline/StanfordCoreNLPServer.java#L297

and `getDocument` checks for `serialized` annotations:

https://github.com/stanfordnlp/CoreNLP/blob/b167dc2dda4bd68cc4a6156e051408e43598f1a2/src/edu/stanford/nlp/pipeline/StanfordCoreNLPServer.java#L345

You would leave `inputSerializer` blank, then build a protobuf:

https://github.com/stanfordnlp/CoreNLP/blob/b167dc2dda4bd68cc4a6156e051408e43598f1a2/src/edu/stanford/nlp/pipeline/CoreNLP.proto#L40

The only annotations actually needed by tregex are the sentence index and the tree:

https://github.com/stanfordnlp/CoreNLP/blob/b167dc2dda4bd68cc4a6156e051408e43598f1a2/src/edu/stanford/nlp/pipeline/StanfordCoreNLPServer.java#L1351
https://github.com/stanfordnlp/CoreNLP/blob/b167dc2dda4bd68cc4a6156e051408e43598f1a2/src/edu/stanford/nlp/pipeline/CoreNLP.proto#L86
https://github.com/stanfordnlp/CoreNLP/blob/b167dc2dda4bd68cc4a6156e051408e43598f1a2/src/edu/stanford/nlp/pipeline/CoreNLP.proto#L89

but probably you'll have to fill in a few more of the annotations in order for the serializer to think it's a valid message.

I should point out that your python script `launch_server`, `query`, etc is literally recreating the server interface in Stanza.  You wouldn't need to use the Stanza models to make use of the Stanza server interface.  Although to be fair, sending trees instead of text to tregex like I just described does not exist Stanza.  Perhaps it would be a useful feature to add.

**Comment by tanloong** [comment]
Thanks for the detailed reply! I really appreciate it!

> You wouldn't need to use the Stanza models to make use of the Stanza server interface. 

Oops. In that case, I am happy to drop my scripts and move to the Stanza interface.

BTW, just now I found that downloading models from inside China [perhaps is not that hard](https://github.com/stanfordnlp/stanza/issues/331#issuecomment-686330766). Using  `stanza.download("en", resources_url="stanford")` works well on my side, without VPN.

**Comment by AngledLuffa** [comment]
I dug into it some more.  There's a bug or an inconsistency in the way the most recent CoreNLP release handles incoming tree annotations.  The result is it currently crashes if you pass in parsed trees instead of raw text, unfortunately.  I will fix that and try to make a new release of CoreNLP within the next week or two.

**Comment by tanloong** [comment]
OK! That's so nice of you!

**Comment by AngledLuffa** [comment]
I'm gonna move issue to the CoreNLP repo, since the bug mentioned is there
